### PR TITLE
fix(workflow): bound event stream observers

### PR DESCRIPTION
**Pull request by kwakayama**
### Motivation

The workflow SSE endpoint allowed unbounded long-lived observers, and aggressive Redis polling amplified held connections into sustained backend load and resource retention.

### Changes

- Bound active SSE observations per handler instance, with configurable aggregate and per-authorized-identity limits. Defaults are 64 total streams and 8 streams per identity.
- Return HTTP 429 before opening a backend observation when either admission limit is exhausted.
- Keep a stream reservation until both iterator and observation asynchronous teardown complete.
- Preserve the Redis observation polling interval at 20 ms because the retained journal contains 64 revisions; a 100 ms gap can lose revisions during a burst and fail the contiguity check.
- Document the response, retry behavior, limit scope, and configuration.

### Testing

- Deno 2.7.7: `src/workflow/http/handler.test.ts` passed 48/48 steps.
- Deno 2.7.7: `src/workflow/backends/redis/index.test.ts` passed 158/158 steps.
- Pinned formatting, lint, and targeted typecheck passed.
- API-reference and public-doc checks passed.
- The repository pre-push hook completed its full formatting, lint, typecheck, generation, and unit-test gate successfully.

## Summary by CodeRabbit

* **New Features**
  * Added configurable limits for active workflow event streams per handler and per authorized identity.
  * Requests exceeding these limits now receive HTTP 429 responses.
  * Stream capacity is automatically released when connections close or are cancelled.

* **Documentation**
  * Documented default limits, custom configuration, and limit behavior.
  * Updated API reference source links.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Automated review: 45/100 — significant concerns, not mergeable as-is

The admission-control design is sound (reserve-before-observe, `release()` wired through every exit path — normal cleanup, abort, setup failure, observation error — and a focused regression test), but CI is currently red on this head and the fix leaves a real fairness gap for a DoS-motivated change.

**Strengths**
- Correctly targets the described amplification vector: caps concurrent SSE observers before opening a backend `observeRunEvents` call, so a rejected request never touches Redis.
- `release()` is idempotent and reachable from every code path (404, 501, thrown error, and all `runEventStream` cleanup routes) — no leaked reservation on the failure branches I traced.
- New test (`handler.test.ts`) exercises the actual boundary condition: 64 admitted, 65th rejected with 429, then a slot freed by cancellation gets reused. That's the right shape of regression test for this bug.

**Concerns**
- **CI is failing on this head** (`ci (lint)`, job `99274357121`): `deno lint` reports `require-yield` on the new test's mock async iterator (`handler.test.ts:724`, `async *[Symbol.asyncIterator]() { await Promise.withResolvers<void>().promise; }` never yields). This is a real bug in the new code, not a flake — please fix and push before this is mergeable.
- **The cap has no per-client fairness.** `MAX_EVENT_STREAMS` is a single global counter with no identity dimension, so one caller can still open all 64 slots itself and lock out every legitimate observer — the fix caps aggregate resource use but doesn't stop the single-attacker DoS the PR description frames as the threat. The repo already has a maintained per-client limiter (`veryfront/middleware`'s `rateLimit`/`authRateLimit`/`MemoryRateLimitStore`, plus `resolveRateLimitClientKey` in `src/security/rate-limit/`) that's the established pattern for exactly this — worth using instead of (or in addition to) the bespoke handler-local counter.
- `MAX_EVENT_STREAMS = 64` is hardcoded rather than exposed via `WorkflowHandlerOptions`, so operators can't tune it for their deployment's connection budget without a fork.
- The cap is per-`createWorkflowHandler`-instance (module-level closure), so it only holds within a single process; in a horizontally scaled deployment the effective ceiling is `64 × instance count` with no coordination. Probably an acceptable trade-off for a first pass, but worth calling out in the PR description as a known limitation rather than left implicit.
- PR description states local test/lint runs couldn't complete due to environment limitations — the lint failure above is exactly the kind of issue that would have caught. Please get a real local/CI run green before requesting review on a security fix.

**Before merge:** fix the `require-yield` lint failure, and at minimum note (ideally address) the lack of per-client fairness — either via the existing middleware rate limiter or a documented follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kwakayama** [comment]
## Codex review

Reviewed SHA: `fc5968ce953eb6a3a1e306bf64c143cb10bd9b30`

Findings, ordered by severity:

1. **High:** `MAX_EVENT_STREAMS` is a single process-local counter, so one authorized identity can occupy all 64 SSE slots and deny other identities. The admission budget needs a per-authorized-identity limit in addition to the aggregate handler limit.
2. **High:** `runEventStream` releases its slot before `iterator.return()` and `observation.close()` finish. Async teardown can therefore exceed the resource bound during rapid cancel/reconnect cycles.
3. **High:** Increasing the Redis observation poll interval to 100 ms while retaining only 64 revisions can make the contiguity check fail during a burst of more than 64 writes in one interval. The stream must retain enough events for the polling strategy, or the interval change must be reverted.
4. **Medium:** The new 429 response and its retry/limit scope are absent from the handler API documentation and advanced workflow guide.
5. **Medium:** The stream limits are hardcoded, preventing operators from tuning the handler budget.
6. **Medium:** The new mock async iterator has no `yield`, so CI `deno lint` fails with `require-yield` at `src/workflow/http/handler.test.ts:724`.

Verification: the PR gate snapshot shows 46 passing, 1 pending, 2 failing checks. The failing jobs are `ci (lint)` and `quality gate (merge)`; CI identifies the `require-yield` error above.

Verdict: REQUEST_CHANGES

Review-Gate:
Reviewer: Codex
Reviewed-SHA: fc5968ce953eb6a3a1e306bf64c143cb10bd9b30
Score: 40/100
Actionable-Findings: 6
Verdict: REQUEST_CHANGES

File: src/workflow/backends/redis/index.ts [hunk 1]
```diff
@@ -59,7 +59,7 @@ const REDIS_STORAGE_SCHEMA_NAMESPACE = `${REDIS_STORAGE_SCHEMA_VERSION}:`;
 const RUN_OBSERVATION_APPROVAL_SCHEMA_VERSION = "approvals-v1";
 const RUN_OBSERVATION_REVISION_FIELD = "__runObservationRevision";
 const RUN_OBSERVATION_STREAM_MAX_LENGTH = 64;
-const RUN_OBSERVATION_POLL_INTERVAL_MS = 20;
+const RUN_OBSERVATION_POLL_INTERVAL_MS = 100;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/workflow/backends/redis/index.ts]: Addressed in `efe302634f5cf6a532ff3511be185c073790d80c`. The unsafe 100 ms polling change was reverted to the existing 20 ms interval because the Redis journal retains 64 revisions and a longer idle gap can lose revisions during a burst. This preserves the revision-contiguity contract; the Redis backend suite passes 158/158.

File: src/workflow/http/handler.ts [hunk 2]
```diff
@@ -57,6 +57,7 @@ export interface WorkflowHandlers {
 }
 
 const DEFAULT_BASE_PATH = "/api/workflows";
+const MAX_EVENT_STREAMS = 64;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/workflow/http/handler.ts]: Addressed in `efe302634f5cf6a532ff3511be185c073790d80c`. The handler now keys reservations by the server-derived identity returned from `authorize`, enforcing both a configurable aggregate limit (default 64) and a configurable per-identity limit (default 8). One identity can no longer consume the entire aggregate budget, while distinct identities can use the remaining capacity. The new handler test covers both fairness and aggregate exhaustion. Pinned handler tests pass 48/48.

File: src/workflow/http/handler.ts [hunk 3]
```diff
@@ -409,12 +424,23 @@ export function createWorkflowHandler(
       }
 
       if (segments.length === 3 && first === "runs" && runId && third === "events") {
-        const observation = await client.observeRunEvents(runId, { signal: request.signal });
-        if (!observation) return problem(`No workflow run ${runId}`, 404);
-        if (!observation.supported) {
-          return problem("Workflow event observation is not supported", 501);
+        const release = reserveEventStream();
+        if (!release) return problem("Too many workflow event streams", 429);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/workflow/http/handler.ts:465]: Addressed in `efe302634f5cf6a532ff3511be185c073790d80c`. The handler API JSDoc, advanced workflow guide, and generated workflow API reference now document the 429 admission response, retry-with-backoff behavior, default/configurable limits, identity scope, and per-process scope. Public-doc and API-reference checks pass.

File: src/workflow/http/handler.ts [hunk 4]
```diff
@@ -230,6 +232,7 @@ function runEventStream(
 
   const cleanup = (): Promise<void> => {
     if (cleanupPromise) return cleanupPromise;
+    release();
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/workflow/http/handler.ts]: Addressed in `efe302634f5cf6a532ff3511be185c073790d80c`. Capacity release now runs in the `Promise.allSettled(...).then(...)` continuation, after both `iterator.return()` and `observation.close()` settle. The new regression test keeps a one-stream handler at 429 while teardown promises are pending, then verifies a replacement is admitted after both resolve. Pinned handler tests pass 48/48.

**Comment by kwakayama** [comment]
**Claude review**

Reviewed SHA: `efe302634f5cf6a532ff3511be185c073790d80c`

### Summary

The change adds aggregate and per-authorized-identity admission control to workflow event-stream SSE responses. Requests beyond either limit receive 429 before backend observation, and reservations remain held through asynchronous teardown.

### Findings

No actionable findings.

Non-blocking suggestions considered:
1. A `Retry-After` header could help clients, but a static value would not reflect the lifetime of an active stream and the documented backoff behavior is more accurate.
2. Admission rejection logging could aid operators, but per-request warning logs would be attacker-amplifiable; existing access metrics/logging can observe the 429 response without adding a new hot-path log.

### Verification

- The handler tests cover aggregate limit/release, per-identity isolation, and delayed teardown.
- The Redis suite passes with the retained 20 ms polling interval.
- API-reference and public-doc checks pass.
- The diff is tightly scoped and the pinned pre-push gate passed.

Verdict: APPROVE

Review-Gate:
Reviewer: Claude
Reviewed-SHA: efe302634f5cf6a532ff3511be185c073790d80c
Score: 95/100
Actionable-Findings: 0
Verdict: APPROVE

**Comment by kwakayama** [comment]
Reviewer: Codex
Reviewed SHA: `efe302634f5cf6a532ff3511be185c073790d80c`

Score: **98/100** — Correctness/completeness 40/40; regression tests/verification 19/20; reliability/security 15/15; repository standards/maintainability 14/15; scope/docs/rollout 10/10.

Findings, highest severity first: **None.** The handler now rejects saturated aggregate or per-identity admission before opening the backend observation, releases setup failures, and retains each reservation until iterator and observation teardown both settle. The implementation is instance-scoped and race-safe under the JavaScript execution model, exposes no new sensitive data, and documents defaults, configuration, 429 behavior, retry guidance, and multi-process scope.

Verification: independently inspected `git diff 93d50c866cbaf147cbd6e75ee09f9561ea4edaa9...efe302634f5cf6a532ff3511be185c073790d80c`, PR body, implementation, tests, and docs. With repository-pinned Deno 2.7.7: handler tests passed 48/48 steps; Redis backend tests passed 158/158 steps; changed TypeScript files passed `deno check`, lint, and format; API-reference and public-doc checks passed; `git diff --check` passed. Gap: the full cross-runtime suite was not rerun locally; at review snapshot the exact head was stable with 35 checks passing, 9 pending, 0 failing, and no unresolved review threads.

Verdict: **APPROVE** — zero blocking or unresolved actionable findings.

Review-Gate:
Reviewer: Codex
Reviewed-SHA: efe302634f5cf6a532ff3511be185c073790d80c
Score: 98/100
Actionable-Findings: 0
Verdict: APPROVE

File: src/workflow/http/handler.ts [hunk 5]
```diff
@@ -57,10 +61,20 @@ export interface WorkflowHandlers {
 }
 
 const DEFAULT_BASE_PATH = "/api/workflows";
+const DEFAULT_MAX_EVENT_STREAMS = 64;
+const DEFAULT_MAX_EVENT_STREAMS_PER_IDENTITY = 8;
 const logger = baseLogger.component("workflow-http");
 
 class WorkflowRequestError extends Error {}
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/workflow/http/handler.ts:68]: Verified against the pinned base and final diff: `WorkflowRequestError` predates this PR and is unchanged. It is a private control-flow sentinel used only to map malformed HTTP request inputs to a 400 response; it is not a public domain error or exported contract. Converting it to a registered `VeryfrontError` would add a public registry slug and change the existing handler behavior without helping this SSE admission fix, so this out-of-scope suggestion is intentionally left unchanged.
